extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPusher.java

```
@@ -88,24 +89,43 @@ public DataSegment push(File inDir, DataSegment segment) throws IOException
         storageDir
     );
 
-    Path outFile = new Path(String.format("%s/%s/index.zip", config.getStorageDirectory(), storageDir));
+    Path outFile = new Path(String.format(
+        "%s/%s/index.zip",
+        config.getStorageDirectory(),
+        UUIDUtils.generateUuid()
+    ));
     FileSystem fs = outFile.getFileSystem(hadoopConfig);
 
     fs.mkdirs(outFile.getParent());
     log.info("Compressing files from[%s] to [%s]", inDir, outFile);
 
     final long size;
+    final DataSegment dataSegment;
     try (FSDataOutputStream out = fs.create(outFile)) {
       size = CompressionUtils.zip(inDir, out);
+      dataSegment = createDescriptorFile(
+          segment.withLoadSpec(makeLoadSpec(outFile))
+                 .withSize(size)
+                 .withBinaryVersion(SegmentUtils.getVersionFromDir(inDir)),
+          outFile.getParent(),
+          fs
+      );
+      Path outDir = new Path(String.format("%s/%s", config.getStorageDirectory(), storageDir));
+      if(!fs.rename(outFile.getParent(), outDir)){
+        throw new IOException(String.format(
+            "Writing conflict detected for [%s], is there some other job running?",
+            segment.getIdentifier()
+        ));
+      }
+    } finally {
+      try{
+        fs.delete(outFile.getParent(), true);
+      } catch (Exception e){
+        // suppress delete exception
```

not sure if we should suppress this, let the task fail if there was an IOException ... that probably means something is wrong with hdfs in general.